protocol/linkpreview/linkpreview.go

```diff
@@ -241,13 +245,105 @@ func (u OpenGraphUnfurler) unfurl() (common.LinkPreview, error) {
 	return preview, nil
 }
 
+type ImageUnfurler struct {
+	url        *neturl.URL
+	logger     *zap.Logger
+	httpClient http.Client
+}
+
+func compressImage(imgBytes []byte) ([]byte, error) {
+	smallest := imgBytes
+
+	img, err := images.DecodeImageData(imgBytes, bytes.NewReader(imgBytes))
+	if err != nil {
+		return nil, err
+	}
+
+	compressed := bytes.NewBuffer([]byte{})
+	err = images.CompressToFileLimits(compressed, img, images.DefaultBounds)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(compressed.Bytes()) < len(smallest) {
+		smallest = compressed.Bytes()
+	}
+
+	if len(smallest) > maxImageSize {
+		return nil, errors.New("image too large")
+	}
+
+	return smallest, nil
+}
+
+func (u ImageUnfurler) unfurl() (common.LinkPreview, error) {
+	preview := newDefaultLinkPreview(u.url)
+	preview.Type = protobuf.UnfurledLink_IMAGE
+
+	headers := map[string]string{"user-agent": headerUserAgent}
+	imgBytes, err := fetchBody(u.logger, u.httpClient, u.url.String(), headers)
+	if err != nil {
+		return preview, err
+	}
+
+	if !isSupportedImage(imgBytes) {
+		return preview, fmt.Errorf("unsupported image type url='%s'", u.url.String())
+	}
+
+	compressedBytes, err := compressImage(imgBytes)
+	if err != nil {
+		return preview, fmt.Errorf("failed to compress image url='%s': %w", u.url.String(), err)
+	}
+
+	width, height, err := images.GetImageDimensions(compressedBytes)
+	if err != nil {
+		return preview, fmt.Errorf("could not get image dimensions url='%s': %w", u.url.String(), err)
+	}
+
+	dataURI, err := images.GetPayloadDataURI(compressedBytes)
+	if err != nil {
+		return preview, fmt.Errorf("could not build data URI url='%s': %w", u.url.String(), err)
+	}
+
+	preview.Thumbnail.Width = width
+	preview.Thumbnail.Height = height
+	preview.Thumbnail.DataURI = dataURI
+
+	return preview, nil
+}
+
 func normalizeHostname(hostname string) string {
 	hostname = strings.ToLower(hostname)
 	re := regexp.MustCompile(`^www\.(.*)$`)
 	return re.ReplaceAllString(hostname, "$1")
 }
 
+// isSupportedImageContentType detects whether a URL ends with one of the
```

Comment:
Wrong function name I believe, `isSupportImageURL`

Reply:
Good catch. I don't know why I assumed the language server would rename the comment too. I thought that was a feature.

Reply:
Done